File: classic/Code/ValidateBST/src/problem/ValidateBST.java

package problem;

public class ValidateBST {
    class TreeNode {
        TreeNode left;
        TreeNode right;
        int val;

        TreeNode(final int x) {
            val = x;
        }
    }

    public boolean isValidBST(final TreeNode root) {
        if (root == null) {
            return false;
        }
        if (root.left != null && root.right != null) {
            if (root.left.val < root.val && root.right.val > root.val) {
                return isValidBST(root.left) && isValidBST(root.right);
            } else {
                return false;
            }
        } else if (root.left == null && root.right != null) {
            if (root.val < root.right.val) {
                return isValidBST(root.right);
            } else {
                return false;
            }
        } else if (root.right == null && root.left != null) {
            if (root.val > root.left.val) {
                return isValidBST(root.left);
            } else {
                return false;
            }
        } else {
            return true;
        }
    }
}

File: classic/Code/ValidateBST.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool valid(TreeNode*root,int MIN,int MAX){
        if(root==NULL){
            return true;
        }
        if(root->val>MIN&&root->val<MAX&&valid(root->left,MIN,root->val)&&valid(root->right,root->val,MAX)){
            return true;
        }else{
            return false;
        }
    }
    bool isValidBST(TreeNode *root) {
        return valid(root,INT_MIN,INT_MAX);
    }
};
